### DISCUSSION TITLE
`next dev` breaks with next version 9.4.4

### DISCUSSION THREAD

**Discussion by Sowed** [vercel/next.js#13777] [Help]
For some reason after running an `npm audit fix`, my version of `next` got upgraded from `9.3.2` to `9.4.4`, and it breaks on running **`next dev`**. All the other npm script (`next build` and `next start`) work like before. I can build and start the project alright, but cannot run `next dev`.  On checking the browser(Chrome, Opera and Firefox), all I can see is a white screen. Testing out another route (`/about`), has the same effect, even with a hard refresh. I get the following server logs...

```bash
$ npm run dev

> my-project-dir@0.1.0 dev A:\path\to\my-project-dir
> next dev

ready - started server on http://localhost:3000
info  - Loaded env from A:\path\to\my-project-dir\.env
> Using external babel configuration
> Location: "A:\path\to\my-project-dir\.babelrc"
event - compiled successfully
event - build page: /
wait  - compiling...
event - compiled successfully
event - build page: /about
wait  - compiling...
event - compiled successfully
```
The only visible change in the console is that the pages are compiled successfully but unlike before with version `9.3.2`, next no longer logs this line after compiling...
```bash
ready - started server on http://localhost:3000
```
I guess it's because of the said [improved log output](https://nextjs.org/blog/next-9-4#improved-log-output). "Sometimes" though I like think more is better. 🤔

Besides the persistent white screen, I get the following error in the browser `console`, which I supsect somewhere along there line, some `undefined` is being passed as json.
```bash
VM693:1 Uncaught SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at Object../node_modules/ally.js/util/platform.js (platform.js:14)
    at __webpack_require__ (bootstrap:862)
    at fn (bootstrap:150)
    at Object../node_modules/ally.js/supports/detect-focus.js (detect-focus.js:19)
    at __webpack_require__ (bootstrap:862)
    at fn (bootstrap:150)
    at Object../node_modules/ally.js/supports/supports.js (supports.js:21)
    at __webpack_require__ (bootstrap:862)
    at fn (bootstrap:150)
```
I am sure the white screen is just an overlay over the pages because I can see the `DOM` in the `Elements` devtools and in the pages' `view source`.

Can someone please help me out. Possibly point me to a breaking change write-up somewhere for this version. Thank you.
And btw, Nextjs is awesome.

**Comment by Timer** [comment]
Please provide a reproducible demo so someone can take a look.

**Reply by danilockthar** [reply]
Same problem here

To reproduce it , go to an existing next project with 9.3 version and put npm audit fix in the console. After the installing put npm run dev and encounter this problems.

**Reply by Sowed** [reply]
> Please provide a reproducible demo so someone can take a look.

 Upgrade to next version `9.4.4` breaks the `next dev`. To reproduce the error 
1. Clone the [`with-typescript-styled-component`](https://github.com/vercel/next.js/tree/canary/examples/with-typescript-styled-components) example.
2.  At the root, add a `.env` file with `NODE_PATH='.'`
3. Run `next dev` 

Installs of `next` versions below 9.4.4 started throwing `http-proxy` audit warnings, and any `npm audit fix` just bumped up the `next` version to `9.4.4`. This took the warning away but broke `dev`. In my case, I figured the problem was with the `NODE_PATH` in the environment variables. Pre `next:9.4.4`, the config below 'just' worked. Tested with version 9.3.2 and 9.3.5.
```js
// .env file at the root of the project

NODE_PATH='.'
```
You will get the reported error in the browser console but nothing in the server log.

The fix for me is to change the `NODE_PATH` to `'/'` 
```js
// .env file at the root of the project

NODE_PATH='/'
```

**Comment by Jaikant** [comment]
I am starting with a fresh install and am running into this issue. The NODE_PATH='/' doesn't help. 
`next build`, and `next start` works all right. The issue is with `next dev`

I tried a fresh install with 9.4.0 and the issue is present.
I then tried 9.3.6 and `yarn dev` **works good** in this version. Which means the issue was introduced in 9.4.0


PS: I am on a mac

**Reply by Sowed** [reply]
@Jaikant, I also had to play cat and mouse with my configs, commenting out env vars and  custom next_configs until I reached to that. 

It's surprising though that the issues shows up in a fresh 9.4.0 install. That version was completely fine with me.

**Reply by Sowed** [reply]
What error are you getting?

**Reply by Jaikant** [reply]
I was getting exactly the same errors you reported on top
```
VM693:1 Uncaught SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at Object../node_modules/ally.js/util/platform.js (platform.js:14)
    at __webpack_require__ (bootstrap:862)
    at fn (bootstrap:150)
    at Object../node_modules/ally.js/supports/detect-focus.js (detect-focus.js:19)
    at __webpack_require__ (bootstrap:862)
    at fn (bootstrap:150)
    at Object../node_modules/ally.js/supports/supports.js (supports.js:21)
    at __webpack_require__ (bootstrap:862)
    at fn (bootstrap:150)

```

**Reply by kartikag01** [reply]
same is happening for me, on Ubuntu. 
Downgrading next version to  9.3.6 works.

 "react": "16.13.1",

**Reply by kartikag01** [reply]
what i investigated is problem comes from `platform` package which was used in `ally.js` and which was used in `@next/react-dev-overlay`

**Comment by Jaikant** [comment]
Exactly the same errors as in the title of the issue.

**Comment by kartikag01** [comment]
The hack is to replace node_modules/ally.js/util/platform.js with https://gist.github.com/chrisvariety/2429354130eac4f6bf364d68e3722a64
It worked for me for version `9.5.2`

**Comment by devalnor** [comment]
This issue seems to be related to relative path set via NODE_PATH env variable.

Since 9.4, Next.js automatically supports the tsconfig.json and jsconfig.json "paths" and "baseUrl"
More info: https://nextjs.org/docs/advanced-features/module-path-aliases

So, I solved the problem by removing the NODE_PATH settings and adding a `jsconfig.json` with

```json
// jsconfig.json 
{
  "compilerOptions": {
    "baseUrl": "."
  }
}
```

**Reply by Fi1osof** [reply]
@devalnor this is not works with ts-node. https://github.com/Microsoft/TypeScript/issues/29272#issuecomment-451651674
If used abs paths in imports e.g. import 'src/modules' insteads import '../../src/module' env NODE_PATH=. ts-node... required.

I have same error. I had to remove NODE_PATH and rewrite code to relative paths. :(((

**Comment by Sowed** [comment]
Every time I `npm audit fix` the project dependencies, I have to play ping pong with `npm run dev` as it fails miserably with `VM693:1 Uncaught SyntaxError: Unexpected token u in JSON at position 0`.

It has now become a strategy of trial and error. I had changed `NODE_PATH='/'` back to `NODE_PATH=.` trying to fix the work flow but it also failed. I couldn't use @devalnor approach of `"baseUrl": "."` as mine it set to  `"baseUrl": "src"` for absolute imports.

I got it working again with the following steps...

- Deleted the `node_modules` and `.next` folders
- Clean installed with `npm install`
- Set  `NODE_PATH='/'`  in the `.env` file
- `npm run dev`
